controller/scheduler-v2/scheduler.go

```diff
@@ -0,0 +1,1190 @@
+		select {
+		case <-s.rectify:
+			s.Rectify()
+		case <-s.stop:
+			log.Info("stopping scheduler loop")
+			close(s.putJobs)
+			return nil
+		case isLeader := <-leaderCh:
+			s.HandleLeaderChange(isLeader)
+		case req := <-s.jobRequests:
+			s.HandleJobRequest(req)
+		case e := <-s.hostEvents:
+			s.HandleHostEvent(e)
+		case e := <-s.jobEvents:
+			s.HandleJobEvent(e)
+		case f := <-s.formationEvents:
+			s.HandleFormationChange(f)
+		case <-s.syncFormations:
+			s.SyncFormations()
+		case <-s.syncJobs:
+			s.SyncJobs()
+		}
+	}
+	return nil
+}
+
+func (s *Scheduler) SyncJobs() {
+	defer s.sendEvent(NewEvent(EventTypeClusterSync, nil, nil))
+
+	log := fnLogger()
+	log.Info("syncing jobs")
+
+	log.Info("getting host list")
+	hosts, err := s.getHosts()
+	if err != nil {
+		log.Error("error getting host list", "err", err)
+		return
+	}
+
+	newJobs := make(Jobs)
+	for _, h := range hosts {
+		hostLog := log.New("host.id", h.ID())
+
+		hostLog.Info(fmt.Sprintf("getting jobs for host %s", h.ID()))
+		activeJobs, err := h.ListJobs()
+		if err != nil {
+			hostLog.Error("error getting jobs list", "err", err)
+			continue
+		}
+		hostLog.Info(fmt.Sprintf("got %d active job(s) for host %s", len(activeJobs), h.ID()))
+
+		for _, job := range activeJobs {
+			s.handleActiveJob(&job)
+			if j, ok := s.jobs[job.Job.ID]; ok {
+				newJobs[j.JobID] = s.jobs[j.JobID]
+			}
+		}
+	}
+
+	for id, j := range s.jobs {
+		if _, ok := newJobs[id]; !ok && j.IsRunning() {
+			s.handleJobStop(j)
+		}
+	}
+
+	s.triggerRectify()
+}
+
+func (s *Scheduler) SyncFormations() {
+	defer s.sendEvent(NewEvent(EventTypeFormationSync, nil, nil))
+
+	log := fnLogger()
+	log.Info("syncing formations")
+
+	log.Info("getting app list")
+	apps, err := s.AppList()
+	if err != nil {
+		log.Error("error getting apps", "err", err)
+		return
+	}
+
+	for _, app := range apps {
+		appLog := log.New("app.id", app.ID)
+
+		fs, err := s.FormationList(app.ID)
+		if err != nil {
+			appLog.Error("error getting formations", "err", err)
+			continue
+		}
+		appLog.Info(fmt.Sprintf("got %d formation(s) for %s app", len(fs), app.Name))
+
+		for _, f := range fs {
+			if _, err := s.updateFormation(f); err != nil {
+				appLog.Error("error updating formation", "release.id", f.ReleaseID, "err", err)
+			}
+		}
+	}
+
+	s.triggerRectify()
+}
+
+func (s *Scheduler) Rectify() {
+	log := fnLogger()
+
+	if !s.isLeader {
+		log.Warn("ignoring rectify as not service leader")
+		return
+	}
+
+	defer s.sendEvent(NewEvent(EventTypeRectify, nil, nil))
+
+	log.Info("starting rectify")
+
+	for key, formation := range s.formations {
+		formationLog := log.New("app.id", key.AppID, "release.id", key.ReleaseID)
+		formationLog.Info("rectifying formation")
+
+		expected := formation.GetProcesses()
+		actual := s.jobs.GetProcesses(key)
+
+		if expected.Equals(actual) {
+			formationLog.Info("formation in correct state", "expected", expected, "actual", actual)
```

Comment:
We should drop this log message (`Rectify` happens every time we get a job event, so this will get logged for each formation each time)